File: AEDS2 - Java/Verde/TP3/JAVA/TP03Q05 - Lista com Alocação Flexível em Java/Alfa.java
class Jogador {
    private int id;
    private String nome;
    private int altura;
    private int peso;
    private String universidade;
    private int anoNascimento;
    private String cidadeNascimento;
    private String estadoNascimento;
    public Jogador(){}
    public Jogador(int id, String nome, int altura, int peso, String universidade, int anoNascimento, String cidadeNascimento, String estadoNascimento){
        setId(id);
        setNome(nome);
        setAltura(altura);
        setPeso(peso);
        setUniversidade(universidade);
        setAnoNascimento(anoNascimento);
        setCidadeNascimento(cidadeNascimento);
        setEstadoNascimento(estadoNascimento);
    }
    public int getId(){
        return id;
    }

    public String getNome() {
        return nome;
    }

    public int getAltura() {
        return altura;
    }

    public int getPeso() {
        return peso;
    }

    public int getAnoNascimento() {
        return anoNascimento;
    }

    public String getCidadeNascimento() {
        return cidadeNascimento;
    }

    public String getEstadoNascimento() {
        return estadoNascimento;
    }

    public String getUniversidade() {
        return universidade;
    }
    public void setId(int id){
        this.id = id;
    }

    public void setAltura(int altura) {
        this.altura = altura;
    }

    public void setAnoNascimento(int anoNascimento) {
        this.anoNascimento = anoNascimento;
    }
    public void setPeso(int peso) {
        this.peso = peso;
    }

    public void setCidadeNascimento(String cidadeNascimento) {
        cidadeNascimento = validar(cidadeNascimento);
        this.cidadeNascimento = cidadeNascimento;
    }

    public void setEstadoNascimento(String estadoNascimento) {
        estadoNascimento = validar(estadoNascimento);
        this.estadoNascimento = estadoNascimento;
    }

    public void setNome(String nome) {
        nome = validar(nome);
        this.nome = nome;
    }
    public void setUniversidade(String universidade) {
        universidade = validar(universidade);
        this.universidade = universidade;
    }
    public String validar(String str){
        if(str.isEmpty())
            str = "nao informado";
        return str;
    }
    public Jogador clone(){
        return new Jogador(getId(),getNome(),getAltura(),getPeso(),getUniversidade(),getAnoNascimento(),getCidadeNascimento(),getEstadoNascimento());
    }


    public void imprimir() {
        System.out.println("## " + getNome() + " ## " +getAltura() + " ## " + getPeso() + " ## " + getAnoNascimento() + " ## " + getUniversidade() + " ## " + getCidadeNascimento() + " ## "+ getEstadoNascimento() +" ##");
    }
    public void ler(String jogador){

        String[] dados = jogador.split(",", 8);
        setId(dados[0]!="" ?Integer.parseInt(dados[0]): -1);
        setNome(dados[1]);
        setAltura(Integer.parseInt(dados[2]));
        setPeso(Integer.parseInt(dados[3]));
        setUniversidade(dados[4]);
        setAnoNascimento(Integer.parseInt(dados[5]));
        setCidadeNascimento(dados[6]);
        setEstadoNascimento(dados[7]);
    }



}



class celula{
    public Jogador elemento;
    public celula prox;

    public  celula(){}
    public  celula(Jogador x){
        this.elemento = x;
        this.prox = null;
    }


}

class listaflexivel{

    private celula primeiro;
    private celula ultimo;

    public listaflexivel(){
        primeiro = new celula();
        ultimo = primeiro;
    }

    public void inserirFim(Jogador x){
        ultimo.prox = new celula(x);
        ultimo = ultimo.prox;

    }

    public void inserirInicio(Jogador x){

        celula tmp = new celula(x);

        tmp.prox = primeiro.prox;

        primeiro.prox = tmp;
        if(primeiro == ultimo){
            ultimo = tmp;
        }
        tmp = null;

    }

    public Jogador removerFim() throws Exception{
        if(primeiro == ultimo)
            throw new Exception("Erro ao remover (vazia!)");
        celula i;
        for(i = primeiro.prox; i.prox != ultimo; i = i.prox ){};
        Jogador elemento = ultimo.elemento;

        ultimo = i;
        i = ultimo.prox = null;
        return elemento;
    }
    public Jogador removerInicio() throws Exception{
        if(primeiro == ultimo)
            throw new Exception("Erro ao remover (vazia!)");

        celula tmp = primeiro;
        primeiro = primeiro.prox;
        Jogador elemento = primeiro.elemento;

        tmp = tmp.prox = null;

        return elemento;

    }
    public void inserir(Jogador x, int pos) throws Exception{
        int tamanho = tamanho();

        if(pos < 0 || pos > tamanho){
            throw new Exception("erro");
        }else if(pos == 0){
            inserirInicio(x);
        }else if(pos == tamanho){
            inserirFim(x);
        }else {

            celula tmp = new celula(x);
            celula i = primeiro;
            for (int j = 0;j < pos; i = i.prox, j++) ;
            tmp.prox = i.prox;
            i.prox = tmp;
            i = tmp = null;
        }
    }
    public Jogador remover(int pos) throws Exception{
        int tamanho = tamanho();
        Jogador elemento = new Jogador();
        if(pos < 0 || pos >= tamanho || primeiro == ultimo) {
            throw new Exception("erro");
        }else if(pos == 0){
            removerInicio();
        }else if(pos == tamanho - 1){
            removerFim();
        }else {

            celula tmp;
            celula i = primeiro;
            for (int j = 0;j < pos; i = i.prox, j++) ;

            tmp = i.prox;
            elemento = tmp.elemento.clone();
            i.prox = tmp.prox;

            i = tmp = null;
        }
        return elemento;

    }


    public void mostrar(){
        int cont = 0;
        for(celula i = primeiro.prox; i != null; i = i.prox, cont++){
            System.out.print("[" + cont + "] ");
            i.elemento.imprimir();
        }
    }

    public int tamanho(){
        int resp = 0;
        for(celula i = primeiro.prox; i != null; i = i.prox, resp++);
        return resp;
    }

    public celula get(int pos){
        celula resp = new celula();
        celula i = primeiro.prox;
        for(int j = 0; j < pos; i = i.prox, j++);
        resp = i;
        i = null;

        return resp;

    }
    public void swap(int pos1, int pos2){
        celula i =primeiro.prox;
        celula j = primeiro.prox;
        celula tmp = new celula();
        for(int k = 0; k < pos1; i=i.prox, k++);
        for(int l = 0; l < pos2; j=j.prox, l++);
        tmp.elemento = i.elemento;
        i.elemento = j.elemento;
        j.elemento = tmp.elemento;

    }



}
public class Alfa {
    private static listaflexivel lista;
    private static Jogador [] jogadores;

    public static void operacao(String data) throws Exception {
        String[] entrada = data.split(" ",3);
        int posicao = 0;
        if (entrada.length > 1)
            posicao = Integer.parseInt(entrada[1]);
        Jogador retorno;
            switch (entrada[0]) {
                case "II":
                    lista.inserirInicio(jogadores[posicao]);
                    break;
                case "IF":
                    lista.inserirFim(jogadores[posicao]);
                    break;
                case "I*":
                    lista.inserir(jogadores[Integer.parseInt(entrada[2])], posicao);
                    break;
                case "RI":
                    retorno = lista.removerInicio();
                    MyIO.println("(R) " + retorno.getNome());
                    break;
                case "RF":
                    retorno = lista.removerFim();
                    MyIO.println("(R) " + retorno.getNome());
                    break;
                case "R*":
                    retorno = lista.remover(Integer.parseInt(entrada[1]));
                    MyIO.println("(R) " + retorno.getNome());
                    break;
        }
    }
    /*public static void criarLog(){
       Arq.openWrite("806347_selecao.txt");
       Arq.println("806347\t" + tempo + "ms" + "\t" + cont + "comparacoes" + "\t" + contMov + "movimentacoes");
       Arq.close();
   }*/
    public static void leArquivo(Jogador[] jogadores){
        Arq.openRead("/tmp/players.csv");
        Arq.readLine(); //remove a primeira linha
        for(int i = 0; i < 3922; i++){
            String data = Arq.readLine();
            jogadores[i] = new Jogador();
            jogadores[i].ler(data);
        }
    }
    //metodo para identificar a parada, FIM
    static boolean parada(String palavra, String parada){
        for(int i = 0; i < parada.length(); i++){
            if(palavra.charAt(i) != parada.charAt(i))
                return true;
        }
        return false;
    }



    public static void main(String[] args) throws Exception {
        int n = 3922;
        int realTam = 0;

        jogadores = new Jogador[n];

        leArquivo(jogadores);
        int posicao;
        String parada = MyIO.readString();
        lista = new listaflexivel();

        while (parada(parada, "FIM")) {
            posicao = Integer.parseInt(parada);
            lista.inserirFim(jogadores[posicao]);
            parada = MyIO.readString();
        }

        int quant = MyIO.readInt();
        for (int i = 0; i < quant; i++) {
            parada = MyIO.readLine();
            operacao(parada);

        }
            lista.mostrar();
    }
}

File: AEDS2 - Java/Verde/TP3/C/TP03Q07 - Fila com Alocação Flexível em C/Alfa.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <time.h>
#include <err.h>

typedef struct Jogador{

    int id;
    char nome[100];
    int altura;
    int peso;
    char universidade[100];
    int anoNascimento;
    char cidadeNascimento[100];
    char estadoNascimento[100];
} Jogador;

char* validar(char str[]){

    if(!str || strcmp(str, "\0") == 0){ //se o campo for vazio, nao informado eh inserido

        return "nao informado";

    }

    return str;
}

void clone(Jogador * jogador, Jogador * copia){

    copia->id = jogador->id;
    strcpy(copia->nome, jogador->nome);
    copia->altura = jogador->altura;
    copia->peso = jogador->peso;
    strcpy(copia->universidade, jogador->universidade);
    copia->anoNascimento = jogador->anoNascimento;
    strcpy(copia->cidadeNascimento, jogador->cidadeNascimento);
    strcpy(copia->estadoNascimento, jogador->estadoNascimento);

}


void imprimir(Jogador * jogador) {
    printf("## %s ## %d ## %d ## %d ## %s ## %s ## %s ##\n", jogador->nome, jogador->altura, jogador->peso,
           jogador->anoNascimento, jogador->universidade, jogador->cidadeNascimento, jogador->estadoNascimento);
}

void ler(char* str, Jogador* jogador){

    char* token = strsep(&str, ",");  // aqui estou separando a string, a linha inteira do arquivo, usando como criterio de separacao a virgula
    if(token != NULL){
        jogador->id = atoi(token);
    }
    else
        jogador->id = -1;
    token = strsep(&str, ",");
    strcpy( jogador->nome, validar(token)); // aqui estou chamando a funcao validar para os atributos do tipo char, se nao houver nada, o nao informado eh inserido

    token = strsep(&str, ",");
    jogador->altura = atoi(token);

    token = strsep(&str, ",");
    jogador->peso = atoi(token);

    token = strsep(&str, ",");
    strcpy( jogador->universidade, validar(token));

    token = strsep(&str, ",");
    jogador->anoNascimento = atoi(token);

    token = strsep(&str, ",");
    strcpy(jogador->cidadeNascimento, validar(token));

    token = strsep(&str, "\n"); // aqui o criterio de separacao foi diferente, como o final da linha acaba em \n , tive que elimina-lo para o print sair corretamente
    strcpy( jogador->estadoNascimento, validar(token));

}


void leArquivo(Jogador* jogadores[]){
    FILE *fp = NULL;
    fp = fopen("players.csv", "r");
    char *dados;

    dados = (char*) malloc(sizeof(char) * 1000);


    fgets(dados, 1000, fp);

    for(int i = 0; i < 3922; i++){ // aqui eu coloquei todos os jogadores em um array de jogadores, em que cada iteracao eu instanciava um jogador novo em cada indice do array
        fgets(dados, 1000, fp);
        jogadores[i] = (Jogador *) malloc(sizeof(Jogador));

        ler(dados, jogadores[i]);
    }
    free(dados);

}
int encontrarMenor(Jogador *array[], int n, int j, int *menor, int *comp) {
    if (j < n) {


        (*comp)++; // Contagem de comparações
        if (strcmp(array[*menor]->nome, array[j]->nome) > 0) {
            (*menor) = j; // Atualiza o índice do menor elemento
        }

        (*menor) = encontrarMenor(array, n, j + 1, menor, comp);
    }
    return *menor;
}




/*void criarLog (double tempo, int comp, int mov){
    FILE *fp = NULL;
    fp = fopen("806347_selecaoRecursiva.txt", "w");
    fprintf(fp, "806347\t %.4fs \t %d comparacoes \t %d movimentacoes", tempo, comp, mov);
    fclose(fp);
}*/
bool parada(char *str){ // aqui eh a funcao que verifica o criterio de parada
    if(str[0] == 'F' && str[1] == 'I' && str[2] == 'M')
        return false;
    else
        return true;
}

/**
/"Classe Lista"
**/

/**
 * fila estatica
 * @author Max do Val Machado
 * @version 2 01/2015
 */

#define MAXTAM    5


typedef struct Celula {
    Jogador* elemento;        // Elemento inserido na celula.
    struct Celula* prox; // Aponta a celula prox.
} Celula;

Celula* novaCelula(Jogador* elemento) {
    Celula* nova = (Celula*) malloc(sizeof(Celula));
    nova->elemento = elemento;
    nova->prox = NULL;
    return nova;
}

//FILA PROPRIAMENTE DITA ========================================================
Celula* primeiro;
Celula* ultimo;


/**
 * Cria uma fila sem elementos (somente no cabeca).
 */
void start () {
    Jogador* jogador = (Jogador*) malloc(sizeof(Jogador));
    primeiro = novaCelula(jogador);
    ultimo = primeiro;
}

/**
 * Remove elemento da fila (politica FIFO).
 * @return Elemento removido.
 */
Jogador* remover() {
    if (primeiro == ultimo) {
        errx(1, "Erro ao remover!");
    }
    Celula* tmp = primeiro;
    primeiro = primeiro->prox;
    ultimo -> prox = primeiro;
    Jogador* resp = primeiro->elemento;
    tmp->prox = NULL;
    free(tmp);
    tmp = NULL;
    return resp;
}

int tamanho()
{
    //mostrar();
    int soma = 0;
    Celula *i;
    for (i = primeiro->prox; i != primeiro && i != NULL; i = i->prox)
        soma++;

    return soma;
}


/**
 * Insere elemento na fila (politica FIFO).
 * @param x int Elemento a inserir.
 */
void inserir(Jogador* x) {

    if(tamanho() == MAXTAM){
        remover();
        ultimo->prox = novaCelula(x);
        ultimo = ultimo->prox;
        ultimo -> prox = primeiro;

    }
    ultimo->prox = novaCelula(x);
    ultimo = ultimo->prox;
}





/**
 * Mostra os elementos separados por espacos.
 */
void mostrar() {
    Celula* i;
    int cont = 0;
    for (i = primeiro->prox; i != primeiro; i = i->prox) {
        printf("[%d] ", cont++);
        imprimir(i ->elemento);
    }
}


int Round(float number)
{
    return (number >= 0) ? (int)(number + 0.5) : (int)(number - 0.5);
}
//tira media das alturas de cada jogador na fila
int mediaAltura(){
    Celula *i;
    int soma = 0;
    for (i = primeiro->prox; i != primeiro && i != NULL; i = i->prox)
        soma += i -> elemento -> altura;

    float media = (float)soma/tamanho();

    return Round(media);

}



void operacao(char* data, Jogador* jogadores []) {
    char *divisao[3];

    divisao[0] = strsep(&data, " ");
    divisao[1] = strsep(&data, " ");

    Jogador *jogador;
    int index = 0;

    if (divisao[1] != NULL)
        index = atoi(divisao[1]);

    if (strcmp("I", divisao[0]) == 0) {
        inserir(jogadores[index]);
        printf("%d\n", mediaAltura());

    } else if (strcmp("R", divisao[0]) == 0) {
        jogador = remover();
        printf("(R) %s\n", jogador->nome);
        //free(jogador);


    }
}






int main(){
    int m = 3922;
    Jogador* jogadores[m];

    leArquivo(jogadores);
    start();


    int posicao;
    char frase [500];
    scanf("%s", frase);

    while(parada(frase)){ // aqui eu tive que ler a entrada como uma string e utilizar do metodo atoi, para transformar a string em um inteiro ->

        posicao = atoi(frase); // -> tive que fazer isso pois era o unico jeito de comparar numeros com o criterio de parada e utilizar como indice do array

        inserir(jogadores[posicao]);
        printf("%d\n", mediaAltura());

        scanf("%s", frase);

    }
    int a;
    int* quanti = (int*) malloc(sizeof(int));
    scanf("%d",quanti);

    int* i = (int*) malloc(sizeof(int));

    for(*i = 0; *i < *quanti; *i = *i + 1){
        char data [50];
        scanf(" %[^\n]", data);
        operacao(data, jogadores);
    }
    free(quanti);
    free(i);
    mostrar();


}

